gensim/models/neighbor_sim.py

```diff
@@ -0,0 +1,204 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+from doc2vec import Doc2Vec
+from word2vec import Word2Vec
+
+from annoy import AnnoyIndex
+
+"""
+Using 3rd party k-NN libraries to improve performance for method DocvecsArray.most_similar().
+
+More information can be found at: https://github.com/piskvorky/gensim/wiki/Ideas-&-Feature-proposals#integrate-a-fast-k-nn-library
+
+Implemented by: Anh Le (anhldbk@gmail.com)
+
+Website: http://bigsonata.com
+
+Release 1: November 2015
+"""
+
+
+class NeighborIndexer(object):
+    """
+    Base class for k-NN libraries
+    """
+
+    def __init__(self, model=None, **kwargs):
+        """
+        Create a new indexer
+        :param model:   Instance of Doc2Vec or Word2Vec
+        :param kwargs:  Additional named parameters
+        """
+        if model is None:
+            raise Exception("Invalid model parameter. Please provide model with an instance of Doc2Vec or Word2Vec")
+        if type(model) is Doc2Vec:
+            self._init_doc2vec_(model)
+        elif type(model) is Word2Vec:
+            self._init_word2vec_(model)
+        else:
+            raise Exception("Invalid model parameter. Please provide model with an instance of Doc2Vec or Word2Vec")
+        self.start_indexing()
+
+    def _init_doc2vec_(self, model):
+        docvecs = model.docvecs
+        docvecs.init_sims()
+        size = len(docvecs.doctag_syn0norm)
+
+        for i in range(size):
+            self.add_item(docvecs.offset2doctag[i], docvecs.doctag_syn0norm[i])
+
+    def _init_word2vec_(self, model):
+        # raise Exception("Not supported at the moment")
+        model.init_sims()
+        size = len(model.syn0norm)
+
+        for i in range(size):
+            self.add_item(model.index2word[i], model.syn0norm[i])
+
+    def get_item(self, label):
+        """
+        Get an item's vector by its label
+        :param label: The label
+        :return: The item's vector if have. Otherwise, returns None
+        """
+        pass
+
+    def add_item(self, label, vector):
+        """
+        Add an item to index
+        :param label: the label of the item (must be unique).
```

Comment:
We don't use Java-style docstrings in gensim. Either describe the params in free text, or structure using the [Google docstring style](https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments).